pkg/utils/utils_virtual.go

```diff
@@ -93,55 +93,44 @@ type OSPNetworkData struct {
 	// Omit Services
 }
 
-func metaData(platformType PlatformType, address string) (netFilter string, macAddress string) {
-	switch platformType {
-	case VirtualOpenStack:
-		metaData, networkData := readOpenstackMetaData()
-		netFilter, macAddress = parseOpenstackMetaData(address, metaData, networkData)
-
-	default:
-		glog.V(2).Infof("Unknown PlatformType: %v", platformType)
-	}
-
-	return
-}
-
-func readOpenstackMetaData() (metaData *OSPMetaData, networkData *OSPNetworkData) {
-	networkData = &OSPNetworkData{}
-
-	rawBytes, err := ioutil.ReadFile(ospNetworkData)
+// ReadOpenstackMetaData reads the meta data from the openstack metadata file
+func ReadOpenstackMetaData() (metaData OSPMetaData, err error) {
+	var f *os.File
+	f, err = os.Open(ospMetaData)
 	if err != nil {
-		glog.Errorf("error reading file %s, %v", ospNetworkData, err)
-		return
+		err = fmt.Errorf("error opening file %s: %w", ospMetaData, err)
 	}
-
-	if err = json.Unmarshal(rawBytes, networkData); err != nil {
-		glog.Errorf("error unmarshalling raw bytes %v from %s", err, ospNetworkData)
-		return
+	defer func() {
+		if e := f.Close(); err == nil && e != nil {
+			err = fmt.Errorf("error closing file %s: %w", ospMetaData, e)
+		}
+	}()
+	if err = json.NewDecoder(f).Decode(&metaData); err != nil {
+		err = fmt.Errorf("error unmarshalling metadata from file %s: %w", ospMetaData, err)
 	}
+	return metaData, err
+}
 
-	metaData = &OSPMetaData{}
-
-	rawBytes, err = ioutil.ReadFile(ospMetaData)
+// ReadOpenstackNetworkData reads the network data from the openstack metadata file
+func ReadOpenstackNetworkData() (networkData OSPNetworkData, err error) {
+	var f *os.File
+	f, err = os.Open(ospNetworkData)
 	if err != nil {
-		glog.Errorf("error reading file %s, %v", ospMetaData, err)
+		err = fmt.Errorf("error opening file %s: %w", ospNetworkData, err)
 		return
 	}
-
-	if err = json.Unmarshal(rawBytes, metaData); err != nil {
-		glog.Errorf("error unmarshalling raw bytes %v from %s", err, ospNetworkData)
-		return
+	defer func() {
+		if e := f.Close(); err == nil && e != nil {
+			err = fmt.Errorf("error closing file %s: %w", ospNetworkData, e)
+		}
+	}()
+	if err = json.NewDecoder(f).Decode(&networkData); err != nil {
+		err = fmt.Errorf("error unmarshalling metadata from file %s: %w", ospNetworkData, err)
 	}
-
-	return
+	return networkData, err
 }
 
-func parseOpenstackMetaData(pciAddr string, metaData *OSPMetaData, networkData *OSPNetworkData) (networkID string, macAddress string) {
-
-	if metaData == nil || networkData == nil {
-		return
-	}
-
+func parseOpenstackMetaData(pciAddr string, metaData OSPMetaData, networkData OSPNetworkData) (networkID string, macAddress string) {
```

Comment:
will the metaData.Devices not be nil if the `ReadOpenstackMetaData` function doesn't return anything or return error?

Reply:
I added a check:

```golang
	if metaData.Devices == nil {
		return "", ""
	}
```

Tell me what you think.

Reply:
maybe just better to use a pointer for the structs and leave the old validation?

```
if metaData == nil || networkData == nil {
		return
	}
```